File: api/discover.php
<?php
session_start();
require_once '../config.php';
requireSession();

$action = $_GET['action'] ?? '';

// ── SCAN ──────────────────────────────────────────────────
if ($action === 'scan') {
    $hostsRaw = callZabbix('host.get', [
        'output'           => ['hostid', 'host', 'name', 'status'],
        'selectInterfaces' => ['ip', 'main'],
        'monitored_hosts'  => 1,
        'limit'            => 1000,
    ]);
    if (!is_array($hostsRaw) || isset($hostsRaw['_zabbix_error'])) jsonOut(['error' => 'Zabbix error'], 500);

    $subnets = [];
    foreach ($hostsRaw as $h) {
        $ip = '';
        foreach (($h['interfaces'] ?? []) as $i) {
            if (($i['main'] ?? 0) == 1) { $ip = $i['ip'] ?? ''; break; }
        }
        $parts = explode('.', $ip);
        $sub = (count($parts) === 4)
            ? $parts[0].'.'.$parts[1].'.'.$parts[2].'.0/24'
            : 'unassigned';

        if (!isset($subnets[$sub])) {
            $subnets[$sub] = ['subnet' => $sub, 'hosts' => []];
        }
        $subnets[$sub]['hosts'][] = [
            'hostid' => $h['hostid'],
            'host'   => $h['host'],
            'name'   => $h['name'],
            'ip'     => $ip,
            'type'   => detectType($h['host'], $h['name']),
        ];
    }

    // Sort hosts alphabetically within each subnet
    // Separate subnets with 2+ hosts from singletons
    $main = []; $small = [];
    foreach ($subnets as $s) {
        usort($s['hosts'], fn($a,$b) => strcmp($a['host'], $b['host']));
        if (count($s['hosts']) >= 2) $main[] = $s;
        else $small = array_merge($small, $s['hosts']);
    }
    usort($main, fn($a,$b) => count($b['hosts']) - count($a['hosts']));

    jsonOut(['subnets' => $main, 'singletons' => $small]);
}

// ── CREATE MAP ────────────────────────────────────────────
elseif ($action === 'create' && $_SERVER['REQUEST_METHOD'] === 'POST') {
    requireOperator();
    $b    = json_decode(file_get_contents('php://input'), true) ?? [];
    $name = trim($b['name'] ?? 'Auto-Discovered Map');
    $selected = $b['subnets'] ?? []; // [{subnet, label, hosts:[{hostid,host,name,ip,type}]}]
    if (empty($selected)) jsonOut(['error' => 'No subnets selected'], 400);

    $db = getDB();

    // Create new layout (is_default=0, never touches the main map)
    $db->prepare("INSERT INTO map_layouts (name, positions, is_default) VALUES (?,?,0)")
       ->execute([$name, '{}']);
    $layoutId = $db->lastInsertId();

    // Layout: subnets in a grid, hosts in a circle per subnet
    // Hub node at center of each cluster connects to every host (star topology)
    $cols  = max(1, (int)ceil(sqrt(count($selected))));
    $gap   = 800;  // px between cluster centers
    $nodes_created = 0;
    $edges = [];
    $nodeStmt = $db->prepare(
        "INSERT INTO map_nodes (id, label, ip, type, x, y, layout_id, zabbix_host_id, status)
         VALUES (?,?,?,?,?,?,?,?,?) ON DUPLICATE KEY UPDATE
         label=VALUES(label),x=VALUES(x),y=VALUES(y),
         layout_id=VALUES(layout_id),zabbix_host_id=VALUES(zabbix_host_id)"
    );

    foreach ($selected as $si => $subnet) {
        $cx = ($si % $cols) * $gap;
        $cy = (int)floor($si / $cols) * $gap;
        $hosts  = $subnet['hosts'];
        $n      = count($hosts);
        $label  = trim($subnet['label'] ?? $subnet['subnet']);
        $radius = max(200, $n * 28);

        // Hub node at cluster center (represents the subnet gateway)
        $hubId = 'hub_'.$layoutId.'_'.$si;
        $nodeStmt->execute([$hubId, $label, $subnet['subnet'], 'switch', $cx, $cy, $layoutId, null, 'ok']);
        $nodes_created++;

        foreach ($hosts as $hi => $host) {
            $angle = (2 * M_PI * $hi) / max(1, $n) - M_PI / 2;
            $x = round($cx + $radius * cos($angle));
            $y = round($cy + $radius * sin($angle));
            $id = 'disc_'.$layoutId.'_'.$nodes_created;
            $nodeStmt->execute([
                $id,
                $host['name'] ?: $host['host'],
                $host['ip'],
                $host['type'],
                $x, $y,
                $layoutId,
                $host['hostid'],
                'ok',
            ]);
            // Star edge: hub → host
            $edges[] = ['from' => $hubId, 'to' => $id];
            $nodes_created++;
        }
    }

    // Store edges in the positions JSON field
    $db->prepare("UPDATE map_layouts SET positions = ? WHERE id = ?")
       ->execute([json_encode(['edges' => $edges]), $layoutId]);

    jsonOut(['ok' => true, 'layout_id' => (int)$layoutId, 'nodes_created' => $nodes_created]);
}

else {
    jsonOut(['error' => 'Unknown action'], 400);
}

// ── TYPE DETECTION ────────────────────────────────────────
function detectType(string $host, string $name): string {
    $s = strtolower($host . ' ' . $name);
    if (preg_match('/fortigate|fortinet/', $s))                       return 'firewall';
    if (preg_match('/\bpalo\b|pa-\d/i', $host.$name))                return 'palo';
    if (preg_match('/\bf5\b|bigip|loadbal/i', $s))                   return 'f5';
    if (preg_match('/\bhsm\b/i', $s))                                return 'hsm';
    if (preg_match('/switch|\bsw[-_]|\btor\b|nexus|catalyst/i', $s)) return 'switch';
    if (preg_match('/\bfw\b|firewall|\bftd\b|\bips\b/i', $s))        return 'firewall';
    if (preg_match('/database|\bdb[-_]|\bsql\b|oracle|mysql|redis/i', $s)) return 'dbserver';
    if (preg_match('/esxi|vmware|vcenter|hyper.v|\bhvn\b|\bolvm\b/i', $s)) return 'infra';
    if (preg_match('/backup|veeam|storeonce|\bsan\b|storage/i', $s)) return 'infra';
    if (preg_match('/router|ag1000|gateway|\bwan\b|\bisp\b/i', $s))  return 'wan';
    return 'server';
}

File: app/routers/discover.py
import json
import math
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import List, Any

from app.deps import get_session, require_operator
from app.services.zabbix_client import call_zabbix
from app.database import execute

router = APIRouter()


def _detect_type(host: str, name: str) -> str:
    s = (host + " " + name).lower()
    import re
    if re.search(r"fortigate|fortinet", s):                        return "firewall"
    if re.search(r"\bpalo\b|pa-\d", s):                           return "palo"
    if re.search(r"\bf5\b|bigip|loadbal", s):                     return "f5"
    if re.search(r"\bhsm\b", s):                                   return "hsm"
    if re.search(r"switch|\bsw[-_]|\btor\b|nexus|catalyst", s):   return "switch"
    if re.search(r"\bfw\b|firewall|\bftd\b|\bips\b", s):          return "firewall"
    if re.search(r"database|\bdb[-_]|\bsql\b|oracle|mysql|redis", s): return "dbserver"
    if re.search(r"esxi|vmware|vcenter|hyper.v|\bhvn\b", s):      return "infra"
    if re.search(r"backup|veeam|storeonce|\bsan\b|storage", s):   return "infra"
    if re.search(r"router|ag1000|gateway|\bwan\b|\bisp\b", s):    return "wan"
    return "server"


@router.get("/scan")
async def scan(session: dict = Depends(get_session)):
    hosts_raw = await call_zabbix("host.get", {
        "output": ["hostid", "host", "name", "status"],
        "selectInterfaces": ["ip", "main"],
        "monitored_hosts": 1,
        "limit": 1000,
    })
    if not isinstance(hosts_raw, list):
        raise HTTPException(500, "Zabbix error")

    subnets: dict = {}
    for h in hosts_raw:
        ip = ""
        for iface in h.get("interfaces", []):
            if str(iface.get("main")) == "1":
                ip = iface.get("ip", "")
                break
        parts = ip.split(".")
        sub = (f"{parts[0]}.{parts[1]}.{parts[2]}.0/24"
               if len(parts) == 4 else "unassigned")
        if sub not in subnets:
            subnets[sub] = {"subnet": sub, "hosts": []}
        subnets[sub]["hosts"].append({
            "hostid": h["hostid"],
            "host":   h["host"],
            "name":   h["name"],
            "ip":     ip,
            "type":   _detect_type(h["host"], h["name"]),
        })

    main_subnets, singletons = [], []
    for s in subnets.values():
        s["hosts"].sort(key=lambda x: x["host"])
        if len(s["hosts"]) >= 2:
            main_subnets.append(s)
        else:
            singletons.extend(s["hosts"])
    main_subnets.sort(key=lambda x: -len(x["hosts"]))

    return {"subnets": main_subnets, "singletons": singletons}


class SubnetSelection(BaseModel):
    subnet: str
    label: str = ""
    hosts: List[Any] = []


class CreateMapBody(BaseModel):
    name: str = "Auto-Discovered Map"
    subnets: List[SubnetSelection]


@router.post("/create")
async def create_map(body: CreateMapBody, session: dict = Depends(require_operator)):
    if not body.subnets:
        raise HTTPException(400, "No subnets selected")

    layout_id = await execute(
        "INSERT INTO map_layouts (name, positions, is_default) VALUES (%s,'{}',0)",
        (body.name,),
    )

    cols = max(1, math.ceil(math.sqrt(len(body.subnets))))
    gap  = 800
    nodes_created = 0
    edges = []

    for si, subnet in enumerate(body.subnets):
        cx = (si % cols) * gap
        cy = (si // cols) * gap
        hosts  = subnet.hosts
        n      = len(hosts)
        label  = subnet.label.strip() or subnet.subnet
        radius = max(200, n * 28)

        hub_id = f"hub_{layout_id}_{si}"
        await execute(
            "INSERT INTO map_nodes (id, label, ip, type, x, y, layout_id, zabbix_host_id, status) "
            "VALUES (%s,%s,%s,'switch',%s,%s,%s,NULL,'ok') "
            "ON DUPLICATE KEY UPDATE label=VALUES(label),x=VALUES(x),y=VALUES(y),layout_id=VALUES(layout_id)",
            (hub_id, label, subnet.subnet, cx, cy, layout_id),
        )
        nodes_created += 1

        for hi, host in enumerate(hosts):
            angle = (2 * math.pi * hi) / max(1, n) - math.pi / 2
            x = round(cx + radius * math.cos(angle))
            y = round(cy + radius * math.sin(angle))
            node_id = f"disc_{layout_id}_{nodes_created}"
            await execute(
                "INSERT INTO map_nodes (id, label, ip, type, x, y, layout_id, zabbix_host_id, status) "
                "VALUES (%s,%s,%s,%s,%s,%s,%s,%s,'ok') "
                "ON DUPLICATE KEY UPDATE label=VALUES(label),x=VALUES(x),y=VALUES(y),"
                "layout_id=VALUES(layout_id),zabbix_host_id=VALUES(zabbix_host_id)",
                (node_id, host.get("name") or host.get("host"), host.get("ip", ""),
                 host.get("type", "server"), x, y, layout_id, host.get("hostid")),
            )
            edges.append({"from": hub_id, "to": node_id})
            nodes_created += 1

    await execute(
        "UPDATE map_layouts SET positions=%s WHERE id=%s",
        (json.dumps({"edges": edges}), layout_id),
    )
    return {"ok": True, "layout_id": layout_id, "nodes_created": nodes_created}
